### DISCUSSION TITLE
Forking: how is it should be handeled currently, can it support this flow?

### DISCUSSION THREAD

**Discussion by EvanCarroll** [semantic-release/semantic-release#2158] [Q&A]
Let's say you take project with any configuration and you fork it, what can be done with this?

This project solicits outside contributions with forking, as do most other projects? If this was hosted on a private instance of GitLab or GitHub I would like the fork to,

* Run tests and lint checking
* Create a build/release which could be downloaded and tested by QA
* Associate these artifacts/tags/releases with the forked version's semver number, for example if you fork and patch master for it to create a totally local copy with an alpha-beta channel.
* To ensure in this schema that alpha/beta were never uploaded or pushed to upstream but ran and were provided *ENTIRELY* on your fork.

**Comment by travi** [comment]
there should be nothing about semantic-release that prevents use on a fork, as long as it is a proper fork. by that, i mean that it is not associated with the original project. assuming you are talking about an npm package, that means renaming the package to one that you have publish rights to.

if that happens, and your ci is configured with the proper secrets to enable semantic-release to follow its normal process, version bumping will happen based on the existing tags, as normal.

**Reply by EvanCarroll** [reply]
In the scope of a system that is built on GitLab and GitHub what I mean by "fork" is a fork of the project such as one that can be done with the "Fork Button", and with the intent of contributing back to Upstream. For reference here is my fork of [`semantic-release`](https://github.com/EvanCarroll/semantic-release).

**Reply by travi** [reply]
a source code fork does not come with rights to publish to the package registry under the same package name, so any versions desired would need to be published under a different name, with appropriate credentials that provide the rights to publish under that name. whether the intent is for contributing upstream or not, the contribution must be accepted by official maintainers before it could be published in any form under the original name. 

am i misunderstanding the situation you are asking about?

**Reply by EvanCarroll** [reply]
I think you understand what I'm wanting to do, but there are some misunderstanding about the environment,

* _"does not come with rights to publish to the package registry under the same package name"_ This is not true as far as I know, because the fork can push to the project-level/package-level [NPM endpoint which does not enforce any naming convention](https://docs.gitlab.com/ee/user/packages/npm_registry/#package-naming-convention). Only the instance-level API has a naming convention.
* Even if you're not utilizing the NPM package registry, you can still have a fork (my definition) make use of `semantic-release` because you could still publish a [GitLab Release](https://docs.gitlab.com/ee/api/releases/), and you could still have a pipeline produce an artifact.
* Yes in order to publish the tag (as a core part of `semantic-release` functionality) the end user would have to log into their forked copy and set the env variables for the GITLAB_TOKEN, but other than that everything should just work -- and even WITHOUT the `GITLAB_TOKEN` the default [`CI_JOB_TOKEN`](https://docs.gitlab.com/ee/ci/jobs/ci_job_token.html) can be used to [create a release](https://docs.gitlab.com/ee/api/releases/index.html#create-a-release).

**Reply by travi** [reply]
> does not come with rights to publish to the package registry under the same package name

what i did not state in my original statement was "to the same registry that the original package is published to". it sounds like you are wanting to publish to an alternative registry (gitlab), effectively shadowing the real package with your fork (to narrow the scope of the discussion, this is not possible with the GH registry because you are required to publish under your scope). 

this is not an approach that we recommend because it is very dependent on the implementation of the alternative registry that you are using. shadowing is explicitly prevented by the official registry, not only because of uniqueness, but also because it is an obvious supply chain attack vector. beyond those risks, it is unclear how whether the registry would return published versions of the original package in the absence of published versions of the shadowed version. if a package with the requested name exists in the local registry, but does not include the versions of the original package, then i would expect the official version would not be found.

regardless, i'm having a hard time understanding what this has to do with `semantic-release`. as long as the CI of the forked project is configured with credentials that allow it to publish to the configured registry (which would need to be changed in your scenario, in the `.npmrc` for example), semantic-release will proceed as normal. in short, if npm can publish in that way, semantic-release can publish in that way.

**Reply by EvanCarroll** [reply]
> regardless, i'm having a hard time understanding what this has to do with semantic-release. as long as the CI of the forked project is configured with credentials that allow it to publish to the configured registry (which would need to be changed in your scenario, in the .npmrc for example), semantic-release will proceed as normal. in short, if npm can publish in that way, semantic-release can publish in that way.

Not really so, `semantic-release` can't really do anything for example (that I can see) because it assumes the git registry is coming from the `package.json` (or `releaserc`) rather than from the CI variable `CI_REPOSITORY_URL`. We could set it down the path of automatically configuring, but none of this is there currently.

The idea here is simply to make it such that a fork of upstream can access to build artifacts and releases that the fork makes. These artifacts are local to the fork. This sounds **tremendously** useful, moreso if it can be achieved automatically (which it seems like it can). But it's not really useful if the fork has to custom-configure it's own package.json and realease.rc. That's a lot of work.

**Reply by travi** [reply]
i think i understand the scenario that you have been describing and i recognize that the plugins for the various VCS services and package registries could be adjusted to make things more flexible to enable such adjustments without committing changes to files like `package.json`, `.npmrc`, or `.releaserc`. 

however, what you describe is an extreme edge case, especially because of the idea of shadowing the original package. the practice is not one we recommend and is not even possible with many registries. that makes this scenario and unsupported workflow for semantic-release. we are a very small team of maintainers with limited time for maintenance. we cannot afford to invest in use cases that do not see broad usage. that applies both the initial investment of implementation, and the long term maintenance of additional complexity for configuration.

semantic-release is targeted at releases of a package by it's maintainers (even if that maintenance includes accepting outside contributions to the official project) to help with automating a repetitive process over a lengthy period of time. forks are separate from those official projects and it is expected that configuration of the forked project is required to make any type of release from the fork. that is likely to include adjustment to the package name, the registry, the repo details, and providing credentials. if the intent of a fork is for upstream contribution, it is typically short lived and often not worth the effort to go to the lengths of all of this configuration in a CI pipeline. it is valid to publish manually using `npm publish` locally in such scenarios.

**Reply by EvanCarroll** [reply]
But I mean what is the workflow if the acceptance of a patch requires a build for QA? Does this project address that?

I've asked [the above question in another discussion](https://github.com/semantic-release/semantic-release/discussions/2160) because I do think VCS-forking is something well worthy of it's own discussion. I don't agree it's "shadowing" either. When you fork a package on a VCS you have two different package repos. It's not useful to think of the VCS fork as "shadowing" so I'm unsure why we would think of the VCS's artifacts (whether or not hosted behind an independent npm registry) as shadowing. A VCS fork is just namespaced differently: it has a different git url behind the repo, and a different npm url behind the registry.

**Reply by travi** [reply]
> I'm unsure why we would think of the VCS's artifacts (whether or not hosted behind an independent npm registry) as shadowing. A VCS fork is just namespaced differently: it has a different git url behind the repo, and a different npm url behind the registry.

i don't understand this statement. packages installed from an npm registry are installed by package name, not by url. there is no namespacing by registry url available when installing by package name. package scopes are the official way to define that separate registries should be used for various packages, but using scopes in that way would require a different name between the original package and the forked package (as i suggested above). since you are suggesting installation of two different packages by the same name from two different registries, i am suggesting that you are indeed wanting to shadow the original package with the forked package.

**Comment by EvanCarroll** [comment]
> since you are suggesting installation of two different packages by the same name from two different registries, i am suggesting that you are indeed wanting to shadow the original package with the forked package.


It's not clear to me why you think that's "shadowing" though. The `@cpanel/` scope exists on public NPM. It's where we publish our public packages. We have an internal repository with Verdaccio that caches *EVERYTHING* on NPM and allows up to security audit it and whitelist modules we want. We have GitLab project called cPanel which is an instance repository that hosts all the packages on GitLab that publish to the instance repository. And when you fork a package GitLab automatically sets up a package repository for it so you can download just packages published to that private repository.

At any point you can make `@cpanel` resolve to public NPM, private Verdaccio, private GitLab, or a package repository. Are all of these "shadowing"? Is it shadowing when you change your DNS resolver? Is it shadowing when you a pin a package with Debian or change the Debian repository? I've never heard of the term "shadowing" here or seen someone present it as a bad thing.

Go to a different source, get a different thing by the same name. Seems pretty typical for a computer network.
